DA/Internal/Desugar.hs

```diff
@@ -97,4 +101,16 @@ class HasToAnyChoice t c r | t c -> r where
 class HasFromAnyChoice t c r | t c -> r where
   _fromAnyChoice : proxy t -> Any -> Optional c
 
-data Implements t i = Implements
+class Implements t i where
+  toInterface : t -> i
+  fromInterface : i -> Optional t
+  toInterfaceContractId : ContractId t -> ContractId i
+  fromInterfaceContractId : ContractId i -> Update (Optional (ContractId t))
+  interfaceTypeRep : t -> TypeRep
```

Comment:
`interfaceTypeRep` is going to be difficult to use as defined, since it doesn't imply the `i` type variable. It also doesn't make much sense to call `interfaceTypeRep` on a template type, you only really need to call it on the interface type.

I think this would be better suited to a separate typeclass:

```
class HasInterfaceTypeRep i where
    interfaceTypeRep :: i -> TypeRep
```

And then create instances of this only for the interface types, not for templates.

Reply:
Honestly I'm not 100% sure how this method is intended to be used. In the current desugaring, we generate `tokenTypeRep : t -> TypeRep`, so it does seem like one would call it on a template type? I think @robin-da added it, perhaps he can help us here?

Regarding the `i` type variable, in an earlier iteration I added a proxy-like argument to avoid `AllowAmbiguousTypes`, but @cocreature said it was okay to add the pragma and call `interfaceTypeRep` with a type application.

Reply:
A separate typeclass sounds good. It is intended to be used on the interface payload as a runtime typecheck, I don’t see a usecase to call it on the template, you already know the template there.

Reply:
Fine to keep as is for this PR though.